### PR TITLE
Add toggle to switch between released and revised versions of a learning object 

### LINKED ISSUE #710: Add Indicator for which version is being viewed for #572
Add some kind of visual representation of state change from released to revised learning objects. 

Proposed solutions: 
* Change the background color of the splash to match the orange color that indicates review stage 

Child of #572 
See mock at https://sketch.cloud/s/MLkjW

### LINKED ISSUE #711: Check for ID or hasRevision flag on Learning Object #572
Check to see if the learning object being loaded into the details page has revisions that can be displayed. 

Child of #572

### LINKED ISSUE #712: Toggle switch implementation for #572
Implement a toggle switch in order for reviewers to toggle between the released and revised versions of a learning object 

Child of #572 
Requires #761

### PR DESCRIPTION
**Pull request by pzalep1**
Screen Shot 2019-03-20 at 11 21 34 AM
Screen Shot 2019-03-21 at 12 53 15 PM
Screen Shot 2019-03-21 at 12 53 24 PM

Contributes to #572 
Resolves #710 
Resolves #711 
Resolves #712 
Needs Cyber4All/learning-object-service#291
Needs Cyber4All/library-service#90

### CODE REVIEW

**Comment by pzalep1** [comment]
I put a work in progress flag on this because the back end is not ready and therefore this is not ready to be deployed.

File: src/app/cube/details/details.component.ts [hunk 1]
```diff
@@ -98,6 +101,26 @@ export class DetailsComponent implements OnInit, OnDestroy {
       }
     });
 
+    this.hasRevisions = true;
+    // delete after dev is done. Just need it to access the UI
```
Comment (nrwinner) [inline src/app/cube/details/details.component.ts]: I see there's a comment here to remove this, can we denote it with a FIXME?

File: src/app/cube/details/details.component.ts [hunk 2]
```diff
@@ -98,6 +101,26 @@ export class DetailsComponent implements OnInit, OnDestroy {
...
+    this.reviewer = true;
+    // this.hasRevisions = this.learningObject.hasRevisions;
+    // this.reviewer = this.auth.hasReviewerAccess();
+
+    if (this.hasRevisions) {
+      this.route.params.pipe(takeUntil(this.isDestroyed$)).subscribe(params => {
+        const learningObjectName = decodeURIComponent(params['learningObjectName']);
+        // will get replaced with function to load revisions
```
Comment (nrwinner) [inline src/app/cube/details/details.component.ts]: Can we denote this with a FIXME?

File: src/app/cube/details/reviewer-panel/reviewer-panel.component.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    <div class="last-revised">
+        Last Revised {{learningObject.date | date: 'shortDate'}}
+      </div>
+      <div class="last-released">
+        Last Released {{learningObject.date | date: 'shortDate'}}
+      </div>
+  </div>
+</div>
+
+<clark-popup *ngIf="showDownloadModal" (closed)="toggleDownloadModal(false)">
```
Comment (nrwinner) [inline src/app/cube/details/reviewer-panel/reviewer-panel.component.html]: Why is this here?

File: src/app/cube/details/reviewer-panel/reviewer-panel.component.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        this.isEditButtonViewable = this.auth.hasEditorAccess();
+      });
+    this.hasDownloadAccess = this.auth.hasReviewerAccess();
+    this.url = this.buildLocation();
+    this.saved = this.cartService.has(this.learningObject);
+    const userName = this.auth.username;
+    this.userIsAuthor = (this.learningObject.author.username === userName);
+  }
+
+  async addToCart(download?: boolean) {
```
Comment (nrwinner) [inline src/app/cube/details/reviewer-panel/reviewer-panel.component.ts]: Can we move the download logic up a component level? That way, both presentation components are "dumb" and the parent component handles the heavy lifting.

File: src/app/cube/details/reviewer-panel/reviewer-panel.component.html (hunk 3 above)
Comment (pzalep1) [reply src/app/cube/details/reviewer-panel/reviewer-panel.component.html]: Moved to the parent component

File: src/app/cube/details/action-panel/action-panel.component.ts [hunk 5]
```diff
@@ -127,10 +111,20 @@ export class ActionPanelComponent implements OnInit, OnDestroy {
     this.addingToLibrary = false;
   }
 
+  async downloadRevised(download?: boolean) {
```
Comment (nvisal1) [inline src/app/cube/details/action-panel/action-panel.component.ts]: This function doesn't need to be async since we aren't awaiting anything.

File: src/app/cube/details/reviewer-panel/reviewer-panel.component.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+})
+export class ReviewerPanelComponent implements OnInit {
+  @Input() releasedDate;
+  @Input() revisedDate;
+  @Output() download: EventEmitter<boolean> = new EventEmitter();
+  constructor() { }
+
+  ngOnInit() {
+  }
+  click(download: boolean) {
```
Comment (nvisal1) [inline src/app/cube/details/reviewer-panel/reviewer-panel.component.ts]: Can you give this function a more descriptive name?

File: src/app/cube/details/version-switch/version-switch.component.html [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<div class="inner">
+  <div *ngIf="hasRevisions === true; then revisions; else noRevisions"></div>
```
Comment (nvisal1) [inline src/app/cube/details/version-switch/version-switch.component.html]: hasRevisions is a boolean. You can check if it is true with `*ngIf="hasRevisions"`.

File: src/app/cube/details/action-panel/action-panel.component.ts (hunk 5 above)
Comment (pzalep1) [reply src/app/cube/details/action-panel/action-panel.component.ts]: Resolved with remove async.

File: src/app/cube/details/reviewer-panel/reviewer-panel.component.ts (hunk 6 above)
Comment (pzalep1) [reply src/app/cube/details/reviewer-panel/reviewer-panel.component.ts]: Resolved with Change click to downloadRevisions.

File: src/app/cube/details/version-switch/version-switch.component.html (hunk 7 above)
Comment (pzalep1) [reply src/app/cube/details/version-switch/version-switch.component.html]: Resolved with Remove === true.

File: src/app/core/cartv2.service.ts [hunk 8]
```diff
@@ -6,6 +6,7 @@ import { saveAs as importedSaveAs } from 'file-saver';
 import { AuthService } from './auth.service';
 import { BehaviorSubject, throwError } from 'rxjs';
 import { catchError, retry } from 'rxjs/operators';
+import { stringify } from '@angular/compiler/src/util';
```
Comment (DonSeannelly) [inline src/app/core/cartv2.service.ts]: Was this an accidental auto import? I don't see it used here

Comment (pzalep1) [reply src/app/core/cartv2.service.ts]: Resolved with remove stringify.

File: src/app/cube/details/action-panel/action-panel.component.html [hunk 9]
```diff
@@ -1,63 +1,68 @@
...
     tipLocation="top">
     <i class="fas fa-exclamation"></i>Review Pending
   </div>
-  <div class="btn-group vertical" [ngClass]="{'review-pending': !isReleased}">
-    <button tip="{{ tips.DOWNLOAD_NOW }}" tipLocation="top" tipDisabled="{{hasDownloadAccess || !loggedin}}" [ngClass]="{'disabled': !auth.user || !hasDownloadAccess || downloading}"
-      class="button good" id="download-button" (click)="addToCart(true)">
-      <span *ngIf="hasDownloadAccess">{{!downloading ? 'Download Now' : 'Downloading'}}
-        <span *ngIf="downloading">
-          <i class="fal fa-spinner-third fa-spin"></i>
+  <div *ngIf="revisedVersion === false; else reviewerPanel">
```
Comment (nrwinner) [inline src/app/cube/details/action-panel/action-panel.component.html]: Keeping in theme with @nvisal1 comment above, this can instead be 
```suggestion
  <div *ngIf="!revisedVersion; else reviewerPanel">
```
Also kudos for using Angular templates here

File: src/app/cube/details/details.component.html [hunk 10]
```diff
@@ -37,8 +37,9 @@
           (deleteRating)="deleteRating($event)" (editResponse)="editResponse($event);" (deleteResponse)="deleteResponse($event)" (reportRating)="reportRating($event)" (respondRating)="submitResponse($event)"></clark-learning-object-ratings>
       </div>
     </div>
-    <div class="sidebar">
-      <cube-details-action-panel [learningObject]="learningObject"></cube-details-action-panel>
+    <div class="sidebar" [ngStyle] = "{'margin-top': reviewer === true ? '-190px' : '-110px'}">
```
Comment (nrwinner) [inline src/app/cube/details/details.component.html]:
```suggestion
    <div class="sidebar" [ngStyle] = "{'margin-top': reviewer ? '-190px' : '-110px'}">
```

File: src/app/cube/details/version-switch/version-switch.component.html [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<div class="inner">
+  <div *ngIf="hasRevisions; then revisions; else noRevisions"></div>
```
Comment (nrwinner) [inline src/app/cube/details/version-switch/version-switch.component.html]: We should consider moving logic and copy specific to this issue outside of the component and keeping this component generic. That way, next time we need a switch somewhere, we have a component that's JUST a switch and we don't have to do anything extra.

File: src/app/cube/details/action-panel/action-panel.component.html (hunk 9 above)
Comment (pzalep1) [reply src/app/cube/details/action-panel/action-panel.component.html]: Resolved with Remove ===.

File: src/app/cube/details/details.component.html (hunk 10 above)
Comment (pzalep1) [reply src/app/cube/details/details.component.html]: Resolved with Remove ===.

File: src/app/cube/details/version-switch/version-switch.component.html (hunk 11 above)
Comment (pzalep1) [reply src/app/cube/details/version-switch/version-switch.component.html]: Resolved with Make switch more generic. It's only a switch now. The only thing that would have to get changed when using it somewhere else is the color of the toggle.

File: src/app/cube/details/details.component.html [hunk 12]
```diff
@@ -37,8 +37,15 @@
           (deleteRating)="deleteRating($event)" (editResponse)="editResponse($event);" (deleteResponse)="deleteResponse($event)" (reportRating)="reportRating($event)" (respondRating)="submitResponse($event)"></clark-learning-object-ratings>
       </div>
     </div>
-    <div class="sidebar">
-      <cube-details-action-panel [learningObject]="learningObject"></cube-details-action-panel>
+    <div class="sidebar" [ngStyle] = "{'margin-top': reviewer ? '-190px' : '-110px'}">
+      <div class="inner"*ngIf="reviewer">
+        <div class="revisions">
```
Comment (EVogel1999) [inline src/app/cube/details/details.component.html]: Couldn't this be just one *ngIf="hasRevisions" statement instead of two?  Maybe an else statement too.

File: src/app/cube/details/details.component.ts [hunk 13]
```diff
@@ -34,6 +34,10 @@ export interface Rating {
 export class DetailsComponent implements OnInit, OnDestroy {
   private isDestroyed$ = new Subject<void>();
   learningObject: LearningObject;
```
Comment (EVogel1999) [inline src/app/cube/details/details.component.ts]: Is there a way to not use this variable?  Currently there are 3 loaded objects at a time and all we need is 2 at most.

File: src/app/cube/details/details.component.ts [hunk 14]
```diff
@@ -168,6 +203,76 @@ export class DetailsComponent implements OnInit, OnDestroy {
       }
       console.log(e);
     }
+
+    if (this.hasRevisions) {
```
Comment (EVogel1999) [inline src/app/cube/details/details.component.ts]: Between getting the released object and the revised one, is there a way to put this logic into a separate method so that this logic isn't repeated too much?

File: src/app/cube/details/details.component.ts (hunk 13 above)
Comment (pzalep1) [reply src/app/cube/details/details.component.ts]: There is actually only two loaded objects at most. The released is always loaded and if the user is a reviewer and there are revisions then the revised copy is loaded as well. The learningObject is what is currently displayed and is loaded with either the released or revised copies depending on the state of the toggle switch. I did this so the html didn't have if statements.

File: src/app/cube/details/details.component.ts (hunk 14 above)
Comment (pzalep1) [reply src/app/cube/details/details.component.ts]: Moved this to a separate function.

File: src/app/cube/details/details.component.html [hunk 15]
```diff
@@ -41,7 +41,7 @@
       <div class="inner"*ngIf="reviewer">
         <div class="revisions">
```
Comment (DonSeannelly) [inline src/app/cube/details/details.component.html]: Should this element be the one to have the `*ngIf="hasRevisions"` since the View Revisions div and the clark-version-switch component should only show if `hasRevisions` is true?

Comment (pzalep1) [reply src/app/cube/details/details.component.html]: Yes, I changed that with change logic commit. I didn't realize I didn't copy and paste in the right div.

Comment (DonSeannelly) [reply src/app/cube/details/details.component.html]: I must have opened the review right before you pushed. I see it now 👍
